### ISSUE TITLE
Refactor the logging system

### ISSUE THREAD

**Issue by luziferius** [autokey/autokey#338]
## Classification:

Enhancement

## Version

AutoKey version: git master/develop
Used GUI (Gtk, Qt, or both): both

## Summary

The current logging system is lacklustre and not really helpful.
Some parts of AutoKey completely lack any logging (like the GTK GUI), and some use different approaches. The Qt5-GUI already uses a hierarchical logger system (but logging is incomplete, nonetheless), but the model, X interface and others use plain logger instances.

To improve the current situation, implement a uniform way to use the logging framework built into Python.
- Add a logging module to the `autokey` package that contains a setup function that initializes the system. For example, re-use this from one of my other side-projects: https://github.com/luziferius/av1transcoder/blob/master/av1transcoder/logger.py
- Add a `logger` instance to each module that can be used to log various actions.
- Use a hierarchical logger, so that sub-loggers inhere the parent configuration.

**Comment by josephj11** [comment]
Is this small enough in scope to be a requirement (blocker) for 0.96? How much work is involved?

I totally agree that the logger needs work. While you're at it, it would be nice to improve the readability of some of the messages. I have to visually parse the log just to see what trigger phrase was typed to start a script. It's pretty hard to read.

**Comment by luziferius** [comment]
It kind of is small enough.

Replacing the existing logging is not that hard, given that I already wrote a working implementation under the GPLv3 that can be copy&pasted.

The most work involves adding logger messages to everything that is currently not logged at positions that make sense. This includes the whole GTK GUI and major parts of the Qt GUI (as that already has logging in parts).

**Comment by BlueDrink9** [comment]
How soon do you think you can get this done? The earlier, the easier it is for me to add bits as I wander through the codebase.

Is there a good way I can start adding logging lines before this is done?

**Comment by luziferius** [comment]
It’s almost done.
You can see the diff against develop here:
https://github.com/autokey/autokey/compare/develop...feature/refactor_logging

TODO:
- - [x] Port existing logging messages.
- - [x] Decide on a uniform per-module setup style.

There are five possible styles:

1. ```python
   import autokey.logger
   logger = autokey.logger.get_logger(__name__)
   ```
2. ```python
   import autokey.logger
   logger = autokey.logger.get_logger(__name__)
   del autokey.logger
   ```

3. ```python
   from autokey.logger import get_logger
   logger = get_logger(__name__)
   ```
4. ```python
   from autokey.logger import get_logger
   logger = get_logger(__name__)
   del get_logger
   ```
5. ```python
   logger = __import__("autokey.logger").logger.get_logger(__name__)
   ```


1 and 2 use absolute imports.
3 and 4 use from style imports.
2 and 4 delete unused references after usage. The get_logger() function is used only once per module (unless the logging of a module can be improved by using multiple logger instances).
5 is a terse one-liner that contains everything, logger module import, logger creation and clean-up.

@BlueDrink9 What’s your opinion on the choices?


Edit:
- There is a behaviour change: AutoKey will now always print logging messages of the INFO severity and above to the standard output in addition to writing them to a file. Previously, it was EITHER printing DEBUG level messages to standard output OR writing INFO level messages to a file (but never both at the same time, giving `--verbose` previously disabled the log file writing).
 The `--verbose` flag continues to set the output level to DEBUG for the standard output only.
- The log messages contain the module path and form a hierarchical system.
- Added a new command line parameter (`--cutelog-integration`) that enables logging to a network socket. Use it to display the verbose DEBUG level logging in [Cutelog](https://github.com/busimus/cutelog).
![Cutelog](https://user-images.githubusercontent.com/2143820/71375722-7deeba80-25bf-11ea-8c80-00958a362ef4.png)

**Comment by josephj11** [comment]
Maybe I should make this should be a separate issue. It's more about what is being logged rather than how logging is done.

Aside from making the system more maintainable, what effect is all this logging work going to have on end users who regularly use it for debugging scripts (and for us when we support other users on the forum/listserv)?

The current logs leave a lot to be desired as to readability and I'm a bit concerned that this could get worse if a lot of new messages are added. We might end up needing more than one logging level.

When does it help to see each separate character as it is detected? From my point of view, it's mostly noise that fills up the log. Maybe this could be reported at a higher log level if it is useful.

E.g.: When a trigger phrase is recognized, I'd like to see the whole phrase - not just character by character:

```
Trigger phrase "my-trigger-phrase" found: Running "my-script"
Hotkey "Super+G" found
    Window filter matched:  Running "my-phrase"
```
or something like that.

I'm not sure what should be reported about window filters, but something seems useful because if the filter isn't matching what the user thinks it should, then it's not obvious that the filter is why unexpected things are happening. If no filter is set, then no messages about it would be needed. Window filter failures are another candidate for a higher logging level.

```
Hotkey "Super+G" found
   Window filter for "thunderbird-script" failed
   Window filter for "kate-script" matched: Running "kate-script"
```
(assuming both of the above scripts/phrases had Super+G defined as their hotkey)

I think all that "Grabbing key" stuff at the start of each log should be eliminated if it hasn't already been. I just want to know if it tries to grab a needed key, etc. and fails for some reason. This could be another higher log level message too.

**Comment by BlueDrink9** [comment]
I love the one-liner. Very nice. Don't even know why you're asking for my input with that there.

I don't see how the import style matters too much, since both styles are used in and around autokey src files

**Comment by BlueDrink9** [comment]
And thanks for fixing the `verbose` behavior, that irritated me

**Comment by BlueDrink9** [comment]
When I'm at my machine, one of my projects uses a logger setup that logs the higher log levels to a separate file in the log folder. Helpful to make it easy to see the big pictures easier.

**Comment by luziferius** [comment]
> I love the one-liner. Very nice. Don't even know why you're asking for my input with that there.

The `__import__` one-liner is rather obscure. I’d like to uniform styles across the project and this is a nice opportunity to do so for the logging system. If you had said that it is a bad idea/style, I’d settle for one of the other choices. But since you like it, I’ll go for it.

> When I'm at my machine, one of my projects uses a logger setup that logs the higher log levels to a separate file in the log folder. Helpful to make it easy to see the big pictures easier.

These lines create a `RotatingFileHandler` …
https://github.com/autokey/autokey/blob/9cf2a91f80546d078e4441ffeefe922d3ac14643/lib/autokey/logger.py#L41-L46
… that is attached to the root logger here and writes the log to file:
https://github.com/autokey/autokey/blob/9cf2a91f80546d078e4441ffeefe922d3ac14643/lib/autokey/logger.py#L53

Adding a new handler that only logs warnings (or worse) is simple: Just duplicate the first block, set a different log file and set another log level, e.g. `WARNING`.


@josephj11

I think the grabbed hotkeys at the beginning are useful sometimes. But I agree on removing the input queue logging on each key press. It should reduce the verbosity without impacting much. Maybe logging it every 10th key press or something like this. It IS helpful sometimes, e.g. here: #95

When you type "asdf" and AutoKey tells you you typed "ASDF", then there is something wrong. Without this, you won’t see why nothing triggers your Phrases at all.

I agree that there are messages that should be enhanced to improve readability.

**Comment by josephj11** [comment]
@luziferius Good point about the individual key logging messages. I think both needs could be served by having one or more additional logging levels (accessible from the CLI with something like `autokey-qt -l 3` or specifying `-verbose` more than once - although I don't like that as much.) That way, if the first level trace doesn't tell you what you need to know, you can rerun with a higher level. But most users won't need to do that most of the time and can access much more readable and concise logs.

**Comment by BlueDrink9** [comment]
Is it possible to add additional error levels? Logging every keypress and hotkey seems essential but annoying. Maybe we just have that as a separate option.

**Comment by luziferius** [comment]
> Is it possible to add additional error levels?

Yes. Just set a positive integer as the logging level, and maybe provide a name constant. Here, 9 may be useful, as that wont be logged when the output is set to log DEBUG level and above.


Just note that the levels have a total ordering. You cant have 2 fully distinct sets of levels. Unless you use multiple instances and attach different handlers to those. That might work.

See also:

https://github.com/python/cpython/blob/ff67612fa3c947c1dcc0a4cb0b55ee472b7edc66/Lib/logging/__init__.py#L80-L98
